File: PDOO-Civitas/CivitasR/lib/jugador_especulador.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Jugador_especulador < Jugador
    
    @@Factor_especulador = 2
    
    public
    def initialize(otro, fianza)
    
      super("", otro)
      @fianza = fianza * 1.0
      
      var = 0
      while(var < @propiedades.length())
        
        self.propiedades.at(var).actualizar_propietario_por_conversion(self)
        var = var + 1
      end
      
    end
    
    public
    def encarcelar(num_casilla_carcel)
      if(debe_ser_encarcelado())
        
        if(self.puedo_gastar(@fianza))
          
          self.paga(@fianza)
          Diario.instance.ocurre_evento("El jugador especulador ha pagado la fianza")
          
        else
          mover_a_casilla(num_casilla_carcel)
          @encarcelado = true

          Diario.instance.ocurre_evento("El jugador especulador ha sido encarcelado")
        end
      end
      
      return @encarcelado
      
    end
    
    public
    def self.get_casas_max
      
      return super * @@Factor_especulador
    end
    
    public
    def self.get_hoteles_max
      
      return super * @@Factor_especulador
    end
    
    public
    def paga_impuesto(cantidad)
      if(@encarcelado || (cantidad/2) < 0)
        return false
      else
        return paga(cantidad/2)
      end
    end
    
    def to_s
      super.to_s + "\n|" + @fianza.to_s
    end
  
  end
end

File: PDOO-Civitas/CivitasJ/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author ramonjtorres
 */
public class JugadorEspeculador extends Jugador{
    
    private static final int FactorEspeculador = 2;
    private float fianza;

    public JugadorEspeculador(Jugador otro, float fianza) {
        super(otro);
        this.fianza = fianza;
        
        for(int i = 0; i < this.propiedades.size(); i++){
        
            this.propiedades.get(i).actualizaPropietarioPorConversion(this);
        }
    }
    
    @Override
    boolean encarcelar(int numCasillaCarcel){
        if(this.debeSerEncarcelado()){
            
            if(this.puedoGastar(fianza)){
            
                this.paga(fianza);
                Diario.getInstance().ocurreEvento("El jugador especulador ha pagado la fianza");
            }
            else{
            
                this.moverACasilla(numCasillaCarcel);
                this.encarcelado = true;
                Diario.getInstance().ocurreEvento("El jugador especulador ha sido encarcelado");
            }
        }
        return this.encarcelado;
    }
    
    @Override
    int getCasasMax(){
        return CasasMax*FactorEspeculador;
    }
    
    @Override
    int getHotelesMax(){
        return HotelesMax*FactorEspeculador;
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
        if(this.encarcelado || (cantidad/2) < 0)
            return false;
        else
            return this.paga(cantidad/2);
    }
    
    @Override
    public String toString(){
         
        return super.toString() + "\nFianza: " + this.fianza; 
    }
}
